src/particle_filter.cpp

```diff
@@ -39,65 +101,103 @@ bool compareSamples(const Sample& a, const Sample& b)
 
 // ----------------------------------------------------------------------------------------------------
 
-void ParticleFilter::resample(unsigned int num_samples)
+void ParticleFilter::resample(std::function<geo::Transform2()> gen_random_pose_function)
 {
     std::vector<Sample>& old_samples = samples_[i_current_];
     std::vector<Sample>& new_samples = samples_[1 - i_current_];
 
     if (old_samples.empty())
         return;
 
-    if (num_samples == 0)
-        num_samples = old_samples.size();
-
-    // Sort all samples (decreasing weight)
-    std::sort(old_samples.begin(), old_samples.end(), compareSamples);
-
-    int k = 0;
-    new_samples.resize(num_samples);
-    for(std::vector<Sample>::const_iterator it = old_samples.begin(); it != old_samples.end(); ++it)
-    {
-        const Sample& old_sample = *it;
+    // Build up cumulative probability table for resampling.
+    // TODO: Replace this with a more efficient procedure
```

Comment:
Should we do this?

Reply:
This was copied from ROS AMCL. So I don't know... It has been there in ROS AMCL for a long time, so I suggest to keep it this way.